### PR TITLE
Blum 3-approximate shortest common superstring algorithm

### PR DESCRIPTION
**Pull request by MaikuZ**
Implementacja algorytmu 3-przybliżonego dla problemu najkrótszego nadsłowa wraz z opisem algorytmu i głównym dowodem otrzymanej jakości rozwiązania.

### CODE REVIEW

File: text/MichałZwonek/opis.tex [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  Niech $c_1,c_2$ będą dwoma cyklami w minimalnym pokryciu cyklowym $C$, z $s_1\in c_1$ oraz $s_2 \in c_2$ Wtedy, $ov(s_1,s_2) \leq w(c_1) + w(c_2)$. 
+\end{lemma}
+\begin{lemma}\label{claim4}
+Nadsłowo $\langle s_{i_0},...,s_{i_{r-1}} \rangle $ jest podsłowem $strings(c, i_0)s_{i_0}$.
+\end{lemma}
+
+\begin{theorem}
+\textbf{TGREEDY} dla zbioru $S = \{s_1, ..., s_m \}$ zwraca nadsłowo $s$ nad $S$ o długości co najwyżej $3\cdot OPT(S)$.
+\end{theorem}
+\begin{proof}
```
Comment (krzysztof-turowski) [inline text/MichałZwonek/opis.tex:137]: For your information: I think that a bit simpler proof you may found in Vazirani (Approximation Algorithms) or in https://www.mimuw.edu.pl/~mucha/teaching/aa2008/ss.pdf - but this one is altogether sufficient.

File: test/test_tgreedy.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import unittest
+
+from superstring import tgreedy
+from common import prefix
```
Comment (krzysztof-turowski) [inline test/test_tgreedy.py:4]: Unused import

File: superstring/tgreedy.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+# first phase of mgreedy, computes the set T
+def _mgreedy_first_phase(words):
+  S = set(words)
+  T = set()
+  while S:
+    best_ov = -1
+    chosen_pair = ('','')
+    for w_1 in S:
+      for w_2 in S:
```
[the remark this answers is not in the document]
Comment (MaikuZ) [reply superstring/tgreedy.py:39]: I believe that is not what we actually need. Combinations(['aab', 'baa'], 2) would only yield [('aab', 'baa')], whereas we need [('aab', 'baa'), ('baa', 'aab')], only the latter pair would result in an overlap of 2. I think what we need in this case is actually product(S, 2) and in the other case permutations(S, 2). What do you think?

File: text/MichałZwonek/opis.tex [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+Najprostszym przykładem może być słowo $s=s_1s_2...s_m$. Jednakże, tak uzyskane nadsłowo będzie niepotrzebnie długie.
+
+Jako, że problem ten ma duże znaczenie praktyczne (w sekwencjonowaniu DNA) naturalnym jest próba znalezienia najkrótszego wspólnego nadsłowa.
+
+Oczywistym podejściem jest następujący algorytm zachłanny. Po kolei łącz słowa mające największą część wspólną nachodzącą na siebie określaną jako \textit{overlap}. Rób tak, aż do uzyskania 
+jednego słowa. Postawiona jest hipoteza, że takie podejście ma współczynnik aproksymacji równy 2, ale przed opublikowaniem pracy przez Blum et al. nie było wiadome, czy taki algorytm jest w ogóle liniową aproksymacją. Okazuje się, że ten algorytm (określać będziemy go poprzez \textbf{GREEDY}) ma współczynnik aproksymacji co najwyżej 4.
+
+Ponadto Blum et al. zaprezentowali alternatywny algorytm, nazywany \textbf{TGREEDY}, dla którego udowodnili, że jest on 3-aproksymacyjny. Sam problem znalezienia najkrótszego wspólnego nadsłowa jest NP trudny.
+
+W pierwszej sekcji przedstawimy podstawowe definicje i notacje, a w kolejnych algorytm \textbf{MGREEDY}, a w następynej sekcji algorytm
```
[the remark this answers is not in the document]
Comment (MaikuZ) [reply text/MichałZwonek/opis.tex:40]: There is a description of greedy in the following part:

"Oczywistym podejściem jest następujący algorytm zachłanny. Po kolei łącz słowa mające największą część wspólną nachodzącą na siebie określaną jako _overlap_. Rób tak, aż do uzyskania 
jednego słowa. Postawiona jest hipoteza, że takie podejście ma współczynnik aproksymacji równy 2, ale przed opublikowaniem pracy przez Blum et al. nie było wiadome, czy taki algorytm jest w ogóle liniową aproksymacją. Okazuje się, że ten algorytm (określać będziemy go poprzez **GREEDY**) ma współczynnik aproksymacji co najwyżej 4."

However, I agree that the mention is easy to miss, so maybe I could rephrase the part you highlighted to something like this:

"W pierwszej sekcji przedstawimy podstawowe definicje i notacje, a w kolejnych algorytm **MGREEDY** będący pewną modyfikacją algorytmu **GREEDY**, taką dla którego analiza aproksymacji jest prostsza. W następnej sekcji 
przedstawimy algorytm **TGREEDY** wykorzystujący algorytm **MGREEDY** by uzyskać pożądany współczynnik aproksymacji. W ostatniej sekcji postaramy się udzielić odpowiedzi na pytanie, który z tych dwóch algorytmów jest lepszy?"

And maybe add something like this to the previous extract:

"[...]Okazuje się, że ten algorytm (określać będziemy go poprzez **GREEDY**) ma współczynnik aproksymacji co najwyżej 4, dowodu tego faktu nie będziemy przybliżać. Sam algorytm jest kluczowy, bo jest punktem startowym dla kolejnych dwóch algorytmów na których w tej pracy się skupimy."

What do you think, does that sound better?

Truth be told, the hardest part in writing the description of the original paper was actually compressing it to 3-4 pages, so that it was still readable, contained the main proof and did not exceed 4 pages. I might have misinterpreted this guideline as a hard upper bound, where it actually was more of an lower bound or a soft upper bound. That is my mistake and I think it might have resulted in poorer reader experience, sorry for that.

File: text/MichałZwonek/opis.tex [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+Wiemy, że $S' \subset S''$, a więc $n' \leq n$. Możemy teraz połączyć te wyniki i dostać, że
+\[n_T \leq n'' \leq n' + w \leq n + w \]
+Łącząc to z faktem, że $|s| \leq n_T + w$ dostajemy, że $|s| \leq n + 2w \leq 3n$.
+\end{proof}
+
+\section*{Uwagi końcowe}
+
+Blum et al. pokazali dodatkowo, że \textbf{GREEDY} osiąga współczynnik aproksymacji równy 4, ale hipoteza czy ten współczynnik można obniżyć do 2 jest wciąż otwarta. 
+Możnaby się pokusić wysunąć hipotezę, że \textbf{TGREEDY} jest lepszy niż \textbf{GREEDY}. Tak nie jest. Są przykłady zbioru $S$ dla których zarówno jeden jak i drugi wypada lepiej.
```
[the remark this answers is not in the document]
Comment (MaikuZ) [reply text/MichałZwonek/opis.tex:166]: Thank you! I figured it would be an interesting addition.

File: text/MichałZwonek/opis.tex (hunk 1 above)
Comment (MaikuZ) [reply text/MichałZwonek/opis.tex:137]: Thank you for the information, I'll take a look at the proof you provided. Thanks!

File: superstring/tgreedy.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (krzysztof-turowski) [reply superstring/tgreedy.py:39]: Yes, you're right, thank you for correction.

File: text/MichałZwonek/opis.tex (hunk 4 above)
[the remark this answers is not in the document]
Comment (krzysztof-turowski) [reply text/MichałZwonek/opis.tex:40]: We may leave it like that. I know that's an advice as an afterthought but in such cases do not hesitate to ask.

I'll try to clean this up when I'll be doing the pass to make the notations of all your notes and codes uniform (as that was taken as a soft bound instead, a bit over my expectations). I guess I'll make these three separate sections: greedy (known to be 4-opt), mgreedy (4-opt, but different), tgreedy (3-opt).

[the remark this answers is not in the document]
Comment (MaikuZ) [reply text/MichałZwonek/opis.tex:40]: Okay, I'll fix the other issues you've already pointed out and then I think I should have the updated PR ready before midnight.

File: text/MichałZwonek/opis.tex [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+\end{lemma}
+
+\begin{proof}
+Udowodnimy lemat dla $first(s)$, dowód dla $last(s)$ jest zupełnie analogiczny.
+
+Początkowo $first(s)=last(s)=s$ dla każdego słowa $s$ w aktualnym zbiorze. Załóżmy, że niezmiennik jest w pewnym momencie naruszony poprzez złączenie dwóch słów 
+$t_1,t_2$ w słowo $t = \langle t_1,t_2\rangle$, czyli, że $t$ ma $first(s)$ jako podsłowo. Wtedy wiemy, że 
+$t = ufirst(s)v$ dla jakichś $u,v$. Skoro $first(s)$ nie może być podsłowem 
+$t_1$ ani $t_2$ (założenie o zbiorze $S$), to $|first(s)| > ov(t_1,t_2)$ oraz 
+$|u| < d(t_1,t_2)$. Czyli, $ov(t_1, s) > ov(t_1, t_2)$, co daje nam sprzeczność z faktem, że $ov(t_1,t_2)$ jest największym.
```
[the remark this answers is not in the document]
Comment (MaikuZ) [reply text/MichałZwonek/opis.tex:76]: Frankly, I don't really see why this is needed. I think I can rephrase it in a neater way, tell me what you think.

Wtedy wiemy, że dla t_1 = ab, t_2 = bc (takiego, że b jest maksymalne), t = ufirst(s)v  oraz t = abc. Skoro first(s) nie może być podsłowem t_1 ani t_2 (założenie o zbiorze S), to wiemy, że first(s) musi zawierać b jako podsłowo i nachodzić przynajmniej jedną literą na a i c. Z tego wynika, że |u| < |a| = d(t_1,t_2), |v| < |c| oraz |first(s)| > ov(t_1, t_2) = |b|. Teraz możemy zauważyć, że ov(t_1, s) >= ov(t_1, first(s)) >= |b| + 1 > ov(t_1, t_2) = |b|, co stoi w sprzeczności z faktem, że ov(t_1, t_2) miał być największą wartość.

I think now it's a bit more graphical this way. How do you think, does it look better like this?

[the remark this answers is not in the document]
Comment (krzysztof-turowski) [reply text/MichałZwonek/opis.tex:76]: OK, that's much cleaner now.

**Comment by MaikuZ** [comment]
I have fixed the issues you highlighted and updated the PR. Additionally to the changes you proposed I have also added a test for empty sets of words.

File: superstring/tgreedy.py [hunk 7]
```diff
@@ -50,6 +41,11 @@ def _mgreedy_first_phase(words):
       T.add(w_1)
   return T
 
+# achieves 4-approximation
+def mgreedy(words):
+  return next(iter(words), [''])[0] + \
```
Comment (MaikuZ) [inline superstring/tgreedy.py:46]: I have decided to make it a bit more complicated than the obvious solutios, so that it works even for words being empty. If I just took a wildcard from the first string it would then crash for words being empty. Also, I wanted to use the dummy character from a string inputted by the user as we do not have any prior knowledge about what kind of characters can be used as the dummy character.

Comment (krzysztof-turowski) [reply superstring/tgreedy.py:46]: That's fine, as this is an educational library, not an industry one, we are allowed to assume certain conventions consistent with the lecture notes - but from a broader perspective I understand your cautiousness about the corner cases.
